tests/core/contracts/test_contract_class_construction.py

```diff
@@ -43,6 +48,51 @@ def test_abi_as_json_string(w3, math_contract_abi, some_address):
     assert math.abi == math_contract_abi
 
 
+def test_contract_init_with_abi_function_name(
+    w3,
+    function_name_tester_contract_abi,
+    function_name_tester_contract,
+):
+    # test `abi` function name does not throw when creating the contract factory
+    contract_factory = w3.eth.contract(abi=function_name_tester_contract_abi)
+
+    # re-instantiate the contract
+    contract = contract_factory(function_name_tester_contract.address)
+
+    # Contract `abi`` function should not override `web3` attribute
+    assert contract.abi == function_name_tester_contract_abi
+
+    with pytest.raises(TypeError):
+        contract.functions.abi[0]
+
+    # assert the `abi` function returns true when called
+    result = contract.functions.abi().call()
```

Comment:
Yikes, this is really not working. I'm trying to figure out where the collisions are occurring. It seems to be intertwined with the `ContractFunctions` class.

Reply:
@kclowes in my debugging, it appears that the `abi` property gets overridden when there is an `abi()` function in the contract. I assume this is because the special characters are ignored, causing both properties to be set with the function as the value. It should instead set `abi` to the ABI, and `abi()` to the function. It doesnt seem like there is a way to force the value to be the actual ABI. From what I see, setting the function to the `abi()` property also sets `abi`.

It is wildly confusing to trace this edge case through the code and perhaps doesnt make sense anyway. It is also an issue with the other class properties like `w3`/`w3()` and `address`/`address()` cases.

I am thinking it might just be better to disallow these specific function names to be used in a contract altogether. Would that be reasonable, especially since I was trying to do this on top of the other overloaded contract functions?

Reply:
Is this just in the `ContractCaller` class? If so, think it would be fine to disallow functions with the `w3`/`abi`/`address` names via the `ContractCaller`, as long as it gets documented well. I wonder if we could raise an error or something if someone tries to access the function name? I don't necessarily think that we should block a contract abi with those method names in there, but I could be convinced if you feel strongly. If we do go that path (not allowing w3/abi/address function names on ContractCaller), would you add an issue so that we can track it?

Reply:
I'm going to timebox this to see if there is anything I can isolate with these issues. I agree these names may not be widely used so it may be better to resolve in a later issue. I'll follow up with a new issue if it cannot be resolved in the next day or two.

Reply:
@kclowes I am tinkering with the idea of using an underscore prefix for the class attributes. `_w3` / `_abi` / `_address` would be the new names. This is working at least to avoid collisions with `w3` / `abi` / `address`. If the contract contains the underscored name, the variable itself is no longer overwritten but the code does not access the function when called. It instead uses the class variable itself and throws an exception.

A couple questions:
1. Is changing these to private variables considered breaking?
2. Is it acceptable to fail initializing the contract if a contract uses this convention with the `_` prefix with the same name?

Reply:
> Is changing these to private variables considered breaking?

As long as they're private, it shouldn't be breaking. A good smoke test would be to see what tests break. I can't remember if those really are private variables or not, but am happy to dig in further if you want a second pair of eyes.

> Is it acceptable to fail initializing the contract if a contract uses this convention with the _ prefix with the same name?

I think that's okay here. Looking into this more, `abi` and `address` are reserved in Solidity, so I don't think that we'll run into a problem with those two, and I think `w3` is enough of an edge case that we probably don't need to worry too much about it. Hopefully someone would raise an issue if they ran into problems.